File: API/1557_Williams_R_Cross_with_200_MA_Filter/CS/WilliamsRCrossWith200MaFilterStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;
using Ecng.Collections;
using Ecng.Serialization;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Williams %R cross strategy with SMA filter.
/// Uses RSI as proxy for Williams %R. Enters on oversold/overbought crossover with SMA trend filter.
/// Exits on percent-based TP/SL.
/// </summary>
public class WilliamsRCrossWith200MaFilterStrategy : Strategy
{
	private readonly StrategyParam<int> _rsiLength;
	private readonly StrategyParam<int> _maLength;
	private readonly StrategyParam<decimal> _oversold;
	private readonly StrategyParam<decimal> _overbought;
	private readonly StrategyParam<decimal> _stopPct;
	private readonly StrategyParam<decimal> _tpPct;
	private readonly StrategyParam<DataType> _candleType;

	private decimal _prevRsi;
	private decimal _entryPrice;
	private decimal _stopDist;

	public int RsiLength { get => _rsiLength.Value; set => _rsiLength.Value = value; }
	public int MaLength { get => _maLength.Value; set => _maLength.Value = value; }
	public decimal Oversold { get => _oversold.Value; set => _oversold.Value = value; }
	public decimal Overbought { get => _overbought.Value; set => _overbought.Value = value; }
	public decimal StopPct { get => _stopPct.Value; set => _stopPct.Value = value; }
	public decimal TpPct { get => _tpPct.Value; set => _tpPct.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public WilliamsRCrossWith200MaFilterStrategy()
	{
		_rsiLength = Param(nameof(RsiLength), 14)
			.SetGreaterThanZero()
			.SetDisplay("RSI Length", "RSI period", "General");

		_maLength = Param(nameof(MaLength), 50)
			.SetGreaterThanZero()
			.SetDisplay("MA Length", "SMA filter period", "General");

		_oversold = Param(nameof(Oversold), 30m)
			.SetDisplay("Oversold", "Oversold level", "General");

		_overbought = Param(nameof(Overbought), 70m)
			.SetDisplay("Overbought", "Overbought level", "General");

		_stopPct = Param(nameof(StopPct), 0.5m)
			.SetGreaterThanZero()
			.SetDisplay("Stop %", "Stop loss percent", "Risk");

		_tpPct = Param(nameof(TpPct), 1m)
			.SetGreaterThanZero()
			.SetDisplay("TP %", "Take profit percent", "Risk");

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	protected override void OnReseted()
	{
		base.OnReseted();
		_prevRsi = 0;
		_entryPrice = 0;
		_stopDist = 0;
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var rsi = new RelativeStrengthIndex { Length = RsiLength };
		var sma = new SimpleMovingAverage { Length = MaLength };

		_prevRsi = 0;
		_entryPrice = 0;
		_stopDist = 0;

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(rsi, sma, ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, sma);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal rsiVal, decimal smaVal)
	{
		if (candle.State != CandleStates.Finished)
			return;

		// TP/SL management
		if (Position > 0 && _entryPrice > 0 && _stopDist > 0)
		{
			if (candle.ClosePrice <= _entryPrice - _stopDist || candle.ClosePrice >= _entryPrice + _stopDist * (TpPct / StopPct))
			{
				SellMarket();
				_entryPrice = 0;
				_stopDist = 0;
			}
		}
		else if (Position < 0 && _entryPrice > 0 && _stopDist > 0)
		{
			if (candle.ClosePrice >= _entryPrice + _stopDist || candle.ClosePrice <= _entryPrice - _stopDist * (TpPct / StopPct))
			{
				BuyMarket();
				_entryPrice = 0;
				_stopDist = 0;
			}
		}

		if (_prevRsi == 0)
		{
			_prevRsi = rsiVal;
			return;
		}

		// Crossover signals with MA filter
		var enterLong = _prevRsi < Oversold && rsiVal >= Oversold && candle.ClosePrice > smaVal;
		var enterShort = _prevRsi > Overbought && rsiVal <= Overbought && candle.ClosePrice < smaVal;

		if (enterLong && Position <= 0)
		{
			BuyMarket();
			_entryPrice = candle.ClosePrice;
			_stopDist = candle.ClosePrice * StopPct / 100m;
		}
		else if (enterShort && Position >= 0)
		{
			SellMarket();
			_entryPrice = candle.ClosePrice;
			_stopDist = candle.ClosePrice * StopPct / 100m;
		}

		_prevRsi = rsiVal;
	}
}

File: API/1501-1600/1557_Williams_R_Cross_with_200_MA_Filter/PY/williams_r_cross_with200_ma_filter_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import RelativeStrengthIndex, SimpleMovingAverage, WilliamsR
from StockSharp.Algo.Strategies import Strategy


class williams_r_cross_with200_ma_filter_strategy(Strategy):
    def __init__(self):
        super(williams_r_cross_with200_ma_filter_strategy, self).__init__()
        self._rsi_length = self.Param("RsiLength", 14) \
            .SetDisplay("RSI Length", "RSI period", "General")
        self._ma_length = self.Param("MaLength", 50) \
            .SetDisplay("MA Length", "SMA filter period", "General")
        self._oversold = self.Param("Oversold", 30) \
            .SetDisplay("Oversold", "Oversold level", "General")
        self._overbought = self.Param("Overbought", 70) \
            .SetDisplay("Overbought", "Overbought level", "General")
        self._stop_pct = self.Param("StopPct", 0.5) \
            .SetDisplay("Stop %", "Stop loss percent", "Risk")
        self._tp_pct = self.Param("TpPct", DataType.TimeFrame(TimeSpan.FromMinutes(5))) \
            .SetDisplay("TP %", "Take profit percent", "Risk")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(5))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._prev_rsi = 0.0
        self._entry_price = 0.0
        self._stop_dist = 0.0

    @property
    def rsi_length(self):
        return self._rsi_length.Value

    @property
    def ma_length(self):
        return self._ma_length.Value

    @property
    def oversold(self):
        return self._oversold.Value

    @property
    def overbought(self):
        return self._overbought.Value

    @property
    def stop_pct(self):
        return self._stop_pct.Value

    @property
    def tp_pct(self):
        return self._tp_pct.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(williams_r_cross_with200_ma_filter_strategy, self).OnReseted()
        self._prev_rsi = 0.0
        self._entry_price = 0.0
        self._stop_dist = 0.0

    def OnStarted2(self, time):
        super(williams_r_cross_with200_ma_filter_strategy, self).OnStarted2(time)
        rsi = RelativeStrengthIndex()
        rsi.Length = self.rsi_length
        sma = SimpleMovingAverage()
        sma.Length = self.ma_length
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(rsi, sma, self.on_process).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, sma)
            self.DrawOwnTrades(area)

    def on_process(self, candle, rsi_val, sma_val):
        if candle.State != CandleStates.Finished:
            return
        # TP/SL management
        if self.Position > 0 and self._entry_price > 0 and self._stop_dist > 0:
            if candle.ClosePrice <= self._entry_price - self._stop_dist or candle.ClosePrice >= self._entry_price + self._stop_dist * (self.tp_pct / self.stop_pct):
                self.SellMarket()
                self._entry_price = 0
                self._stop_dist = 0
        elif self.Position < 0 and self._entry_price > 0 and self._stop_dist > 0:
            if candle.ClosePrice >= self._entry_price + self._stop_dist or candle.ClosePrice <= self._entry_price - self._stop_dist * (self.tp_pct / self.stop_pct):
                self.BuyMarket()
                self._entry_price = 0
                self._stop_dist = 0
        if self._prev_rsi == 0:
            self._prev_rsi = rsi_val
            return
        # Crossover signals with MA filter
        enter_long = self._prev_rsi < self.oversold and rsi_val >= self.oversold and candle.ClosePrice > sma_val
        enter_short = self._prev_rsi > self.overbought and rsi_val <= self.overbought and candle.ClosePrice < sma_val
        if enter_long and self.Position <= 0:
            self.BuyMarket()
            self._entry_price = candle.ClosePrice
            self._stop_dist = candle.ClosePrice * self.stop_pct / 100
        elif enter_short and self.Position >= 0:
            self.SellMarket()
            self._entry_price = candle.ClosePrice
            self._stop_dist = candle.ClosePrice * self.stop_pct / 100
        self._prev_rsi = rsi_val

    def CreateClone(self):
        return williams_r_cross_with200_ma_filter_strategy()
